python/replicate/experiment.py

```diff
@@ -144,80 +129,44 @@ def checkpoint(
-            id=random_hash(),
-            created=datetime.datetime.utcnow(),
+        checkpoint = self._project._daemon().create_checkpoint(
+            experiment=self,
             path=path,
             step=step,
             metrics=metrics,
             primary_metric=primary_metric_dict,
+            quiet=quiet,
         )
-        if not quiet:
-            if path is None:
-                console.info("Creating checkpoint {}".format(checkpoint.short_id()))
-            else:
-                console.info(
-                    "Creating checkpoint {}: copying '{}' to '{}'...".format(
-                        checkpoint.short_id(),
-                        checkpoint.path,
-                        self._project._get_repository().root_url(),
-                    )
-                )
-
-        errors = checkpoint.validate()
-        if errors:
-            for error in errors:
-                console.error("Not saving checkpoint: " + error)
-            return checkpoint
-
-        checkpoint._experiment = self
-
-        # Upload files before writing metadata so if it is cancelled, there isn't metadata pointing at non-existent data
-        if checkpoint.path is not None:
-            tar_path = checkpoint._repository_tar_path()
-            repository = self._project._get_repository()
-            repository.put_path_tar(self._project.directory, tar_path, checkpoint.path)
-
         self.checkpoints.append(checkpoint)
-        self.save()
-
-        if self._heartbeat is not None:
-            self._heartbeat.ensure_running()
-
+        self.save(quiet=quiet)
         return checkpoint
 
-    def save(self):
+    def save(self, quiet: bool):
         """
         Save this experiment's metadata to repository.
         """
-        repository = self._project._get_repository()
-        repository.put(
-            self._metadata_path(),
-            json.dumps(self.to_json(), indent=2, cls=CustomJSONEncoder),
-        )
+        self._project._daemon().save_experiment(self)
+        return
 
     def refresh(self):
         """
         Update this experiment with the latest data from the repository.
         """
-        repository = self._project._get_repository()
-        data = json.loads(
-            repository.get("metadata/experiments/{}.json".format(self.id))
-        )
-        fields = experiment_fields_from_json(data)
-        for k, v in fields.items():
-            setattr(self, k, v)
+        exp = self._project._daemon().get_experiment(experiment_id_prefix=self.id,)
+        self.created = exp.created
+        self.user = exp.user
+        self.host = exp.host
+        self.command = exp.command
+        self.config = exp.config
+        self.path = exp.path
+        self.params = exp.params
+        self.python_version = exp.python_version
+        self.python_packages = exp.python_packages
+        self.replicate_version = exp.replicate_version
+        self.checkpoints = exp.checkpoints
```

Comment:
It'd probably be neater and more robust to iterate with this: https://docs.python.org/3/library/dataclasses.html#dataclasses.fields

Reply:
Good point, fixing.